### ISSUE TITLE
Slices and views with asContiguous copies the whole parent tensor data

### ISSUE THREAD

**Issue by edubart** [mratsim/Arraymancer#121]
```Nim
import arraymancer

let a = ones[int](1000,2)
let b = a[0,_]
echo b.data.len

let c = a.unsafeSlice(0,_).asContiguous()
echo c.data.len
```

Outputs:
```
2000
2000
```

This is because of the deep seq copy, this is causing slowdown in axis iterators (used by reduce) and anywhere copying slices.

**Comment by edubart** [comment]
Fixing slicing and asContiguous could be done, but assignment to views could not, would require overloading `=`, suppose this:

```Nim
import arraymancer

let a = ones[int](1000,2)
let c = a.unsafeSlice(0,_)
let d = c
echo d.data.len
```

To fix the above overloading `=` would be required, but that is buggy in Nim atm.

**Comment by mratsim** [comment]
We could use term-rewriting macros to fuse unsafeSlice(...).asContiguous() in one operation.

**Comment by edubart** [comment]
The problem here is that the output should have been 
```
2
2
```

This means that slicing is unnecessarily copying the whole parent tensor data.

**Comment by mratsim** [comment]
I've introduced a "clone" for Tensors that will "compress" slices.
```Nim
import ../src/arraymancer

let a = ones[int](1000,2)
let b = a[0,_]
echo b.data.len

let c = a.unsafeSlice(0,_).asContiguous()
echo c.data.len

let d = a.unsafeSlice(0,_).clone()
echo d.data.len
```
```
2000
2000
2
```


I've also made changes to asContiguous:
- It now compresses like clone
- Except that if it's used in method chaining it will instead use an "unsafe" version.

By using Nim [move {call} optimization](https://nim-lang.org/docs/manual.html#ast-based-overloading-move-optimization), I'd like to introduce that only the first proc call copies, all the other are unsafe. Consequently if the first proc is unsafe, the whole chain is unsafe. hat avoids unsafeProcFoo.unsafeProcBar.unsafeProcBaz
